### DISCUSSION TITLE
Why is FileSystem->isAdminDir response backwards?

### DISCUSSION THREAD

**Discussion by mc12345678** [zencart/zencart#5810] [Q&A]
Quoted code below is from Version 1.5.7; however, the identified methods in Version 1.5.8 are the same at this time.

https://github.com/zencart/zencart/blob/f9c31a44a6a6b10e978f2129a2333bf64339f9dd/includes/classes/FileSystem.php#L66-L72

The above code evaluates `$filePath`. As the method is named, it appears that it is to provide a boolean `true` if `$filePath` is an admin directory. Based on review of the first few lines of the code, it appears that the method of detection/determination is in comparing the provided path to the path of the admin (`DIR_FS_ADMIN`). It would seem that if the path included `DIR_FS_ADMIN`, then the provided path represented the admin directory, if it did not include `DIR_FS_ADMIN` then the provided path is *not* in the admin but somewhere else.

So, if that logic is correct, then it would seem that when reaching:
````php
        $test = str_replace(DIR_FS_ADMIN, '', $filePath);
````
That if `$filePath` pointed to an admin directory folder or file, that `$test` would be everything in `$filePath` that is after `DIR_FS_ADMIN`.

Thus because the provided `$filePath` pointed to an admin directory and was "modified" to obtain `$test` that when `$test != $filePath` should indicate that `$filePath` DOES point to an admin directory, but:
````php
        if ($test != $filePath) return false;
````

Which indicates/points to idea that `->isAdminDir($filePath)` being false actually means that `$filePath` did point to an admin directory rather than how it appears to be implemented.

A similar discussion holds for the method `isCatalogDir($filePath)`: https://github.com/zencart/zencart/blob/f9c31a44a6a6b10e978f2129a2333bf64339f9dd/includes/classes/FileSystem.php#L74-L82

**Comment by drbyte** [comment]
Your point has merit.

I wonder what breaking-changes occur by fixing it...

**Comment by mc12345678** [comment]
New issue discovered by this so called clean up of the issues section. This particular thread did not begin here as a discussion, it was initiated in the issues section. As such there is referential history that is not captured here. As in solutions to the above. 

For more information or history see issue #4432.

**Comment by mc12345678** [comment]
Seeing as there is still no interest in this or the solutions proposed in the original issue, why don't you just close this? It's obvious that intent is no other file is to use this helper function. Otherwise it would have been fixed already.

**Comment by piloujp** [comment]
Well... It is kind of incredible to see some obvious error like this let to rot and still there after few years and Zen Cart versions!
As @mc12345678 clearly pointed out, logic in `isAdminDir` function is reversed, anybody with basic coding skill can see that, no?
Looking at all the related issues raised, the short answer was show us a case where it does not work. Which in my opinion is just a way to avoid dealing with the problem and even kind of stupid.
Of course there isn't, because this function does not do what it should and nobody is going to use it. Some may try but won't be successful.
Well, I found an exception actually in POSM, where it is used to link to documentation in few pages. And it works!
Why is that?
As it is not so rare in ZC, an error is hiding another one.
This function is called only two times in ZC, in `../includes/init_includes/init_observer.php` and `../includes/modules/extra_functions.php` like this:
`$context = (new FileSystem)->isAdminDir(__DIR__) ? 'admin' : 'catalog';`
The PHP constant `__DIR__` refers to the actual script path, which is in catalog side... This test is useless and will always give the same result. In this case it should be 'catalog', but as function `isAdminDir` logic is reversed it is always 'admin'. In POSM, file to be loaded, `auto.posm_help.php` is in admin side, luckily... or not. If you wanted to load some file from catalog side, it will be impossible with this code.

Honestly, those how decide what to do with issues and PR should, at least from time to time, question themselves. Looking at issues like this where time is wasted on discussions and excuses to do nothing instead of acting is kind of sad.

**Comment by scottcwilson** [comment]
This function works as expected in Zen Cart 2.1.0.   Test case: 

Create a file called `test.php` at the top level of your Zen Cart installation. 

```
<?php
use Zencart\FileSystem\FileSystem;
require 'includes/application_top.php';
$context = (new FileSystem)->isAdminDir(__DIR__) ? 'admin' : 'catalog';
echo $context;
```

Run YOURSITE/test.php.  Shows "catalog"
Copy this file to your admin folder.
Run YOURSITE/admin/test.php.  You will have to login of course.  Then it shows "admin"

**Comment by piloujp** [comment]
I did try this and you get the good result, but it is just in appearance.
With the file in root of cart, `DIR_FS_ADMIN` is not defined and function return false without even going through 'reversed logic' loop.
When the file is in admin folder, `DIR_FS_ADMIN` is something like `'My/local/path/admin/'` and `__DIR__` is like this  `'My/local/path/admin'`.
In this case, `$test` is always equal to `__DIR__` (`$filePath` parameter in fact) because `str_replace('My/local/path/admin/', '', 'My/local/path/admin')` does nothing.
Unfortunately, due to security, I was not able to test the file in some subfolders.
`isAdminDir` function returns a boolean. It is why, depending on the bugs, you still have a good chance to get the right result, but it will certainly not be in 100% of cases.

**Comment by scottcwilson** [comment]
> With the file in root of cart, DIR_FS_ADMIN is not defined and function return false without even going through 'reversed logic' loop.

This is the correct behavior.   If you're not in the admin folder, you're not in the admin folder. 

> Unfortunately, due to security, I was not able to test the file in some subfolders.

Executables in the admin folder are at the top level. 

The current logic is working correctly for any reasonable use case I can think of.

**Reply by piloujp** [reply]
> Executables in the admin folder are at the top level.

If it is the same on catalog side, there should be no problem. But any call from a subfolder and this code will fail (tested).

> The current logic is working correctly for any reasonable use case I can think of.

I am not sure of what logic you speak of. Where is the logic in this weird loop that is completely useless? Code for the function could be shortened to this:
```
    public function isAdminDir(string $filePath): bool
    {
        if (!defined('DIR_FS_ADMIN')) {
            return false;
        }
        return true;
    }
```

**Reply by mc12345678** [reply]
Ok, so it's unreasonable to use the feature in an admin/includes/modules file?

What about a catalog file (modules, functions, init_includes, etc) included/required to an admin application that expects an admin response because run from the admin? 

Or what about from the admin feeding the function the directory associated with some other file verified to exist but unknown to be in the admin to cause something other than an `admin`/`catalog` response?

Or is it that this class method is intended to identify that the file having this decision is or is not a file run from the root of the admin? If so, it should be renamed to something like `isAdminRootDir` and some amount of sugar left behind.

**Comment by scottcwilson** [comment]
> What about a catalog file (modules, functions, init_includes, etc) included/required to an admin application that expects an admin response because run from the admin?

If you really need to check to see if you're running in admin context, you should use `if (defined('IS_ADMIN_FLAG'))` - this is used close to 200 times in the storefront.  I wasn't even aware of the `isAdminDir` function until I saw this report.   Is it cruft?  Should it be deleted?  Jeez, I don't know - I have about a million more pressing issues to worry about.

**Reply by mc12345678** [reply]
>  > What about a catalog file (modules, functions, init_includes, etc) included/required to an admin application that expects an admin response because run from the admin?
> 
> If you really need to check to see if you're running in admin context, you should use `if (defined('IS_ADMIN_FLAG'))` - this is used close to 200 times in the storefront.  I wasn't even aware of the `isAdminDir` function until I saw this report.   Is it cruft?  Should it be deleted?  Jeez, I don't know - I have about a million more pressing issues to worry about.  

Oh, so you advocate code development and revision to stick with the old way instead of using new code and functionality. 

Why then have so many thousands of instances of `array()`, `sizeof()`, function parameters, etc been uselessly modified to `[]`, `count()`, and use type casting, respectively? 

Besides it's not like 1) a solution hasn't been provided/suggested for over 4 years or 2) that you're the only one presumably still working on this software, unless you've run the others away.

It's unfortunate @piloujp that even as you've tried to offer some guidance, the results remain the same. Maybe someone who would actually like the core software to benefit the community, development and growth will chime in further, or perhaps as previously and publicly threatened I'll no longer be allowed to access this under my current user profile. 

As to not being aware of this function? 

Sir, I kindly throw the flag. 

I'll give you that you may have forgotten, but it is you who applied the `Ready to Review` label between 3 and 4 years ago on pull request #4435, responded to issue/discussion #4434 that referenced #4432 in the text and had a callout from #4435, transferred the issue (#4432) to a discussion and closed the issue otherwise in preparation of a version release. 

Did you really do all that without consideration/review of the associated topic(s)? Do you still operate that way?

**Comment by piloujp** [comment]
It seems I revived some old wounds...
I found this because I try to understand how things load in ZC and I must say looking at code like this function makes it very hard.
It is barely used anyway in ZC, but in two files, `../includes/init_includes/init_observers.php` and `../includes/modules/extra_functions.php`.
I took a look at the observer loading (especially for plugins) and found out it is included in `config.core.php` files, one in catalog and one in admin. I found other logic error there too, but it is not the point.
The problem is that I found some bad logic in `isAdminDir` function that link to another one deeper in ZC core. In the end, I fail to understand why having high standards when making PRs if ZC core continues to use broken code for years...?

Well, I am back to the end of my first post here:

> Honestly, those how decide what to do with issues and PR should, at least from time to time, question themselves. Looking at issues like this where time is wasted on discussions and excuses to do nothing instead of acting is kind of sad.

**Comment by lat9** [comment]
With trepidation, I'll weigh in.

1. The `isAdminDir` method (and its sibling `isCatalogDir`) is inspecting a specific _directory_ for admin vs. catalog location, rather than the overall _environment_ in which a script is running -- detectable via a check of `IS_ADMIN_FLAG`.  It's unclear to me why these methods were added originally in support of encapsulated plugins.  Perhaps @zcwilt can shed some light?
3. I believe that the `FileSystem->isAdminDir` method is being misused by both `init_observers.php` and `extra_functions.php` and I agree with the changes proposed by @mc12345678 to those two files in [this](https://github.com/mc12345678/zc-v1-series/commit/092de5a6ff339b152bcfb698d73fcb05cd125079#diff-53a540f90421bc1ac35765ade5f87ecc3147d5e1ab8e790fb3238f0c536acaab) commit.  Essentially, both of those scripts are shared between the admin and catalog and the check should result in a load of admin-directory files when in the admin or catalog-directory files otherwise.

**Reply by piloujp** [reply]
Completely agree with that!
Just be aware that the commit you linked missed a problem in function `isAdminDir` with the trailing slash in `DIR_FS_ADMIN`. As long as `$filePath` is a subdirectory of admin, no problem, but if `$filePath` (which in actual code does not have a trailing slash) is actually the admin directory path, the string replacement and test will fail.
The same goes on catalog side.
A solution would be to use `realpath(DIR_FS_ADMIN)`.

**Reply by mc12345678** [reply]
I note mostly from perspective of original effort:

1) I asked the question of (intended/desired) purpose before proposing a solution.
2) The method(s) existed before zc_plugins catalog features were considered available. 
3) For years before the question there has been discussion of eliminating the `admin` directory, though expected environment conditions haven't been reported. E.g. would a constant even exist for the admin directory?
4) method(s) appear to offer useful functionality possibly requiring calling both with `&&` or `||` or even associated `not`. Consider the two functions evaluating if a user is logged in or without an account. 
5) if there ×is× a need/desire for identifying specifically that a file/call is in the root of the admin, then that functionality should remain with either an alternate method or alternate response provided. I prefer an alternate method to accomplish that unique evaluation. 
6) with all that is being expanded by addition of the zc_plugins directory, having a file class that includes evaluation possibilities within that context is helpful in handling background repetitive work. Unfortunately to maximize its use the methods need to function as expected. 
7) I don't yet know what differences may need to be implemented associated with Cron jobs. One may need to define something or other action to support successful use of the admin lookup. 
8) lastly, in absence of further guidance I suggested changes to not negatively impact the nonexistent catalog functionality, while offering operation of what appeared necessary to evaluate any provided.

**Reply by mc12345678** [reply]
> Completely agree with that! Just be aware that the commit you linked missed a problem in function `isAdminDir` with the trailing slash in `DIR_FS_ADMIN`. As long as `$filePath` is a subdirectory of admin, no problem, but if `$filePath` (which in actual code does not have a trailing slash) is actually the admin directory path, the string replacement and test will fail. The same goes on catalog side. A solution would be to use `realpath(DIR_FS_ADMIN)`.

Have a little concern about making a change in `isAdminDir` to meet this particular case in passing only the `dirname` or `realpath` of an admin file to the method. The importance of either of those is that the final slash is absent in the comparison. 

The reason I have concern is in the use of the related method: `getRelativeDir`. In this case, the result of `getRelativeDir` when passed just the text up to the last slash is that the return currently is the full path itself (not relative). If the equivalent replacement is done (either `dirname` or `realpath`, then the result is an empty string (not necessarily a problem), but if still adjusted by either of those and passed a file or folder that includes a slash after the admin's directory, then use of `dirname` or `realpath` will leave a slash at the beginning which is not relative... :/ 

Now, I bring up `dirname` as compared to `realpath` to offer the possibility of `DIR_FS_ADMIN`/`DIR_FS_CATALOG` to
1) contain some other text/reference not specifically in the file system and 2) to not support folder resolution that occurs through `realpath` as compared to `dirname`. In this way also (in my opinion) discovery of the actual admin directory is in a way reduced if there are "unexpected" modifications to either of the two constants... 

Going to continue editing to consider application of `dirname` or `realpath` and possibly changing the variable name within respective functions... In most cases the variable is `$filepath` instead of `$dirpath` where some past use has been the path of the folder not the file...

**Reply by piloujp** [reply]
> The importance of either of those is that the final slash is absent in the comparison.

Yes, and that is all there is in the original context (isAdminDir and isCatalogDir functions). Like that test is done properly. Nothing more to add...
The rest of your post is a completely different discussion based on your own assumption. I hope you realize that makes it hard for busy people to get your point and follow you... I took time to read and understand it, but it is a new subject, and I can see why one would just fly over and go back to its previous activities.

**Reply by mc12345678** [reply]
See #7032 for code changes that fix the issues.

**Reply by mc12345678** [reply]
> Just be aware that the commit you linked missed a problem in function `isAdminDir` with the trailing slash in `DIR_FS_ADMIN`. 

> As long as `$filePath` is a subdirectory of admin, no problem, but if `$filePath` (which in actual code does not have a trailing slash) is actually the admin directory path, the string replacement and test will fail. The same goes on catalog side. 

> A solution would be to use `realpath(DIR_FS_ADMIN)`.

I've separated the above into different "paragraphs" as they each summarize an idea.

Yes, `DIR_FS_ADMIN` and `DIR_FS_CATALOG` each are expected to end with a slash e.g. `/x/y/z/`

The path provided to `isAdminDir` or `isCatalogDir` could be any of the following formats (file extensions `.xxx` and `.xxy` represent any extension):
- `/x/y/z`
- `/x/y/z/` Handled
- `/x/y/z/file1.xxx` Handled
- `/x/y/z/subdir1` Handled
- `/x/y/z/subdir1/` Handled
- `/x/y/z/subdir1/file2.xxy` Handled
- `/x/y/w` Handled

Where `Handled` is stated, the previous code proposal processed sufficiently.

So yes, if `DIR_FS_ADMIN = '/x/y/z/'` and `$filePath = '/x/y/z'` then in using `$test = str_replace(DIR_FS_ADMIN, '', $filePath);` will cause `$test === $filePath`.
This is because DIR_FS_ADMIN is not in $filePath...

Thing is, and all of this is to lead to where I then incorrectly attempted to use the above identified code in an alternate way. I didn't want to use `realpath()` on `DIR_FS_ADMIN` in case there had been some unusual way for it to be modified to an undesirable value, I didn't want to give some sort of additional vector to take action, so I attempted to use `dirname()` though it too didn't give a desired result. When I saw the first problem report of the issue (your report) in using `dirname()`, I went back to investigate. What I found out though was unless I misunderstood the intended use of `realpath()`, it wouldn't help in the above situation.

`realpath('/x/y/z/') === '/x/y/z/'`
`realpath('/x/y/z') === '/x/y/z'`
`realpath('/x/y/z/') !== '/x/y/z'`
`realpath('/x/y/z') !== '/x/y/z/'`

Meaning, if `$filePath = '/x/y/z'` then using `realpath()` in place of my failed use of `dirname` doesn't "fix" the issue in my original code.

I fix that in this commit: https://github.com/zencart/zencart/commit/ca00fa9b6864235d783369086f9d5e556188e99b 

The primary result/benefit is an admin plugin can load a catalog related file, trying first to load it from within the plugin's catalog directory, but if not found then to try to load it in the catalog itself. Further, this assists plugin transition from legacy to plugin structure without unusual and undocumented autoloader path parameter modification, e.g. `filePath`, `classPath`, `loadFile`.

**Reply by lat9** [reply]
Your examples hold for *nix operating systems, but fail on Windows.  I know that Zen Cart doesn't "support" Windows servers, but I (and others) use an XAMPP/Windows based test environment, so your assertions don't hold.
